website/www/site/content/en/blog/apache-beam-flink-and-kubernetes.md

```diff
@@ -0,0 +1,355 @@
+
+<img class="center-block"
+src="/images/blog/apache-beam-flink-and-kubernetes/job-metrics.png"
+alt="Flink Job Metrics">
+
+<img class="center-block"
+src="/images/blog/apache-beam-flink-and-kubernetes/autoscaling-metrics.png"
+alt="Flink Job Autoscaling Metrics">
+
+
+Streaming latency based on timestamp watermark: not only the numbers of events in Kafka as backlogs, it is also important to know the time latency for streaming end-to-end so that we can define and monitor SLA. The latency is defined as the time taken for the streaming processing, starting from ingestion timestamp, to the timestamp sending to streaming endpoint.  A watermark is the last processed event’s time. With the watermark, we are tracking P100 latency. We track each event’s stream latency, so that we can understand each Kafka topic/partition or Flink job pipeline issue. This is an example of each event stream and its latency:
+
+<img class="center-block"
+src="/images/blog/apache-beam-flink-and-kubernetes/watermark-metrics.png"
+alt="Apache Beam Watermark Metrics">
+
+
+## Flink Open Source UI
+
+We use and extend Apache Flink Dashboard UI so that we can closely monitor jobs and tasks, such as the  checkpoint duration, size, and failure. One important extension we did is a job history page so that we can look back at a job's start/update timeline and details so that we can debug issues.
+
+<img class="center-block"
+src="/images/blog/apache-beam-flink-and-kubernetes/flink-checkpoint-ui.png"
+alt="Flink Checkpoint UI">
+
+
+
+## Dashboards and Alerting for Backlog and Latency
+
+We have ~30k jobs, and we want to closely monitor and alert those jobs which run into abnormal states, so that we can quickly intervene. We have created dashboards for each application so that we can show the top latency list, and alert them based on thresholds. This is one example of the timestamp based latency dashboard for one application. We can set the alerting if the latency is larger than a threshold, such as 10 minutes, for a certain time continuously:
+
+
+<img class="center-block"
+src="/images/blog/apache-beam-flink-and-kubernetes/latency-graph.png"
+alt="Latency Graph">
+
+
+Another example of backlogs based dashboards:
+
+
+<img class="center-block"
+src="/images/blog/apache-beam-flink-and-kubernetes/backlog-graph.png"
+alt="Backlog Graph">
+
+
+Alerting is threshold based with frequent metrics checking. If a threshold is met, and continues for a certain amount of times, we will alert our internal Slack channels, or PagerDuty for immediate attention. We tune the alerting so that the accuracy is high.
```

Comment:
```suggestion
The alerts are based on thresholds, and we frequently check metrics. If a threshold is met and continues for a certain amount of times, we alert our internal Slack channels or PagerDuty for immediate attention. We tune the alerting so that the accuracy is high.
```